### ISSUE TITLE
Exhaustive check of `None`s with multiple optional variables fails

### ISSUE THREAD

**Issue by PolarPayne** [python/mypy#5855]
This is a bug, or at least looks like a bug, but might just be a limitation of mypy (in which case it should be documented). Following code is the most minimal repro I could find, I also couldn't find any related issues or documentation.

```python3
from typing import Callable, Optional, Tuple


def f(a: Optional[int], b: Optional[str]) -> Tuple[int, str]:
    if a is None and b is None:
        return 0, ""
    elif a is not None and b is None:
        return a, ""
    elif a is None and b is not None:
        return 0, ""
    # making this into an elif will work
    # elif a is not None and b is not None:
    # as will an assertion
    # assert a is not None and b is not None
    return a, b
```

The given code fails with `error: Incompatible return value type (got "Tuple[Optional[int], Optional[str]]", expected "Tuple[int, str]")`, although all possible `None`s are checked exhaustively and `a` and `b` are no longer `None` on the `return a, b` line.

Mypy version: `mypy 0.641`
Python version: `Python 3.6.6`

**Comment by gvanrossum** [comment]
Yeah, mypy doesn't realize that the exclusion of the three conditions is equivalent to "neither a nor b is None".

Given the structure of the mypy subsystem that keeps track of such exclusions I expect it will be complicated to fix this; it's better to avoid that by restructuring your code, e.g.
```py
if a is None:
    if b is None:
        return 0, ""
    return 0, b
if b is None:
    return a, ""
return a, b
```

**Comment by PolarPayne** [comment]
Interesting, and as an added bonus your example is also a bit clearer to read (at least in my code).
Thank you.
